framework/src/base/FEProblem.C

```diff
@@ -555,13 +560,23 @@ void FEProblem::initialSetup()
       computeUserObjects(EXEC_TIMESTEP_BEGIN);
       computeUserObjects(EXEC_LINEAR);
     }
-    Moose::setup_perf_log.pop("Initial computeUserObjects()","Setup");
-  }
 
-  _nl.initialSetupBCs();
-  _nl.initialSetupKernels();
+    // Perform an initial residual evaluation to set save-in variables
+    // Note: time kernels need not to be evaluated on initial because their contribution should be just zero.
+    //       They actually cannot be evaluated because we set the time step size to zero on initial.
+    if (_nl.hasSaveIn())
+    {
+      _nl.computeResidual(*_nl.sys().rhs, Moose::KT_NONTIME);
+      // We want to call post user object evaluation one more time in case they depends on the save-in variables.
+      computeUserObjects(EXEC_INITIAL, UserObjectWarehouse::POST_AUX);
```

Comment:
@friedmud 
1) residual evaluation definitely possibly need UOs, which can be dictated in `FEProblem::computeResidualType()`, so we need to put residual evaluation on INITIAL at the end of all UOs on INITIAL are computed;
2) There could be cases where a UO depends on the save-in variables, like in the failing test `combined:nodal_area.Hex20` on INITIAL. We need an extra call to resolve this dependency. 
3) During normal transient, the UOs depending on save-in variables are typically on timestep_end. But if they are on linear, we could end up with cyclic dependency. These UOs will not behave like what we expected. LINEAR will not reproduce timestep_end.
4) `TimeKernel` cannot be evaluated on INITIAL currently because we will have 1/dt, where dt = 0.